src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncRewriter.AsyncIteratorRewriter.cs

```diff
@@ -632,10 +632,31 @@ private void GenerateIAsyncEnumerableImplementation_GetAsyncEnumerator()
                 GenerateIteratorGetEnumerator(IAsyncEnumerableOfElementType_GetEnumerator, ref managedThreadId, initialState: StateMachineStates.InitialAsyncIteratorStateMachine);
             }
 
-            protected override BoundStatement GetExtraResetForIteratorGetEnumerator()
+            protected override void GenerateResetInstance(ArrayBuilder<BoundStatement> builder, int initialState)
             {
-                // disposeMode = false;
-                return F.Assignment(F.InstanceField(_disposeModeField), F.Literal(false));
+                // this.state = {initialState};
+                // this.builder = System.Runtime.CompilerServices.AsyncVoidMethodBuilder.Create();
+                // this.disposeMode = false;
+
+                builder.Add(
+                    // this.state = {initialState};
+                    F.Assignment(F.Field(F.This(), stateField), F.Literal(initialState)));
+
+                // this.builder = System.Runtime.CompilerServices.AsyncVoidMethodBuilder.Create();
+                AsyncMethodBuilderMemberCollection methodScopeAsyncMethodBuilderMemberCollection;
+                bool found = AsyncMethodBuilderMemberCollection.TryCreate(F, method, typeMap: null, out methodScopeAsyncMethodBuilderMemberCollection);
+                Debug.Assert(found);
```

Comment:
The base class has an `_asyncMethodBuilderMemberCollection` field. Consider using that here rather than calling `TryCreate()` again which does extra work (and might produce duplicate warnings?). Same comment for `GenerateConstructor()` above.